src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/ClosedRangeInt32FrozenDictionary.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Frozen
+{
+    /// <summary>Provides a frozen dictionary to use when the key is an <see cref="int"/> and the default comparer is used and the item count is small.</summary>
+    /// <typeparam name="TValue">The type of the values in the dictionary.</typeparam>
+    /// <remarks>
+    /// No hashing here, just a straight-up linear scan through the items.
+    /// </remarks>
+    internal sealed class ClosedRangeInt32FrozenDictionary<TValue> : FrozenDictionary<int, TValue>
```

Comment:
These specialized `Int32` dictionaries and sets could be extended to work with any type that implements `IBinaryInteger<TSelf>`.

```csharp
internal sealed class ClosedRangeIntegerFrozenDictionary<TKey, TValue> where TKey : IBinaryInteger<TKey>
{
    private readonly TKey[] _keys;
    private readonly TValue[] _values;
    private readonly TKey _min;

    // assumes the keys are sorted
    internal ClosedRangeIntegerFrozenDictionary(TKey[] keys, TValue[] values)
        : base(EqualityComparer<int>.Default)
    {
        Debug.Assert(keys.Length != 0);

        _keys = keys;
        _values = values;
        _min = keys[0];
    }

    private protected override TKey[] KeysCore => _keys;
    private protected override TValue[] ValuesCore => _values;
    private protected override Enumerator GetEnumeratorCore() => new Enumerator(_keys, _values);
    private protected override TKey CountCore => _keys. Length;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private protected override ref readonly TValue GetValueRefOrNullRefCore(TKey key)
    {
        int intKey = int.CreateTruncating(key - _min);
        return ref (intKey < _values.Length) ? ref _values[intKey] : ref Unsafe.NullRef<TValue>();
    }
}
```

Reply:
We can't introduce this constraint in `TKey` at the moment.

Reply:
So, we'd have to do typechecking instead? That might be worth it, but it would be a lot more code for sure.
```csharp
if (typeof(TKey) == typeof(int))
    intKey = ((int)(object) key) - ((int)(object) _min)
if (typeof(TKey) == typeof(long))
    intKey = (int)(((long)(object) key) - ((long)(object) _min))
// etc.
```

Reply:
@teo-tsirpanis Why can't we have this constraint? The following is a working implementation using this approach. The code in FrozenSet/FrozenDictionary which selects the specific implementation to instantiate can be conditionalized on NET7_0_OR_GREATER such that on .NET 7+ you get optimized implementations for all integer types, but on older runtimes you only get optimized int32 support.

```csharp
namespace System.Collections.Frozen
{
    internal sealed class ClosedRangeIntFrozenSet<TInt> : FrozenSetInternalBase<TInt, ClosedRangeIntFrozenSet<TInt>.GSW> where TInt : IBinaryInteger<TInt>
    {
        private readonly TInt[] _items;
        private readonly TInt _min;

        internal ClosedRangeIntFrozenSet(TInt[] entries)
            : base(EqualityComparer<TInt>.Default)
        {
            Debug.Assert(entries.Length != 0);

            _items = entries;
            _min = _items[0];
        }

        private protected override TInt[] ItemsCore => _items;
        private protected override Enumerator GetEnumeratorCore() => new Enumerator(_items);
        private protected override int CountCore => _items.Length;

        private protected override int FindItemIndex(TInt item)
        {
            var delta = int.CreateTruncating(item - _min);
            if ((uint)delta < (uint)_items.Length)
            {
                return delta;
            }

            return -1;
        }

        internal struct GSW : IGenericSpecializedWrapper
        {
            private ClosedRangeIntFrozenSet<TInt> _set;
            public void Store(FrozenSet<TInt> set) => _set = (ClosedRangeIntFrozenSet<TInt>)set;

            public int Count => _set.Count;
            public IEqualityComparer<TInt> Comparer => _set.Comparer;
            public int FindItemIndex(TInt item) => _set.FindItemIndex(item);
            public Enumerator GetEnumerator() => _set.GetEnumerator();
        }
    }
}
```

Reply:
`ClosedRangeIntFrozenSet`'s generic parameter has a constraint to implement `IBinaryInteger<TInt>`. `FrozenSet.ToFrozenSet`'s generic parameter has no such constraints. There's no way to write `if (T is IBinaryInteger<T>) { }` at the moment. We would need dotnet/csharplang#6308.

Reply:
Ah, well, I just did:

```
if (typeof(T) == typeof(long))
...
```

throw in a few of those in the code, and you've covered the major cases. This is an optimization so it doesn't need to be complete, just to be an incremental improvement.

Do you see any issues if I rework this PR, and the existing int32-specific collections to use this approach? Probably with just built-in explicit support for short/ushort/int/uint/long/ulong?

Reply:
Looks OK to me. Also add a `struct` constraint to `TInt`; it might make some AOT compilers generate less code.

Reply:
Done!